packages/remix-react/data.ts

```diff
@@ -36,16 +36,18 @@ export async function fetchData(
   let url = new URL(request.url);
   url.searchParams.set("_data", routeId);
 
-  let init: RequestInit | undefined;
+  let init: RequestInit = { signal: request.signal };
 
   if (request.method !== "GET") {
-    init = {
-      method: request.method,
-      body: await request.formData(),
-      signal: request.signal,
-    };
-  } else {
-    init = { signal: request.signal };
+    init.method = request.method;
+
+    let contentType = request.headers.get("Content-Type");
+    init.body =
+      // Check between word boundaries instead of startsWith() due to the last
+      // paragraph of https://httpwg.org/specs/rfc9110.html#field.content-type
+      contentType && /\bapplication\/x-www-form-urlencoded\b/.test(contentType)
+        ? new URLSearchParams(await request.text())
+        : await request.formData();
```

Comment:
Conditionally set `URLSearchParams` or `FormData` as the body and let `Content-Type` inherit automatically